Traduction de useReducer (#533)
reactjs/fr.react.dev: src/content/reference/react/useReducer.md

      <button onClick={() => {
        dispatch({ type: 'incremented_age' })
      }}>

>         Incrémenter l'âge

Reviewer: ```suggestion
        Incrémenter l’âge
```

</Sandpack>

> `useReducer` est très similaire à [`useState`](/reference/react/useState), mais il vous permet de déplacer la logique de mise à jour de l'état des gestionnaires d'événements vers une seule fonction à l'extérieur de votre composant. Apprenez-en davantage [pour choisir entre `useState` et `useReducer`](/learn/extracting-state-logic-into-a-reducer#comparing-usestate-and-usereducer).

Reviewer: ```suggestion
`useReducer` est très similaire à [`useState`](/reference/react/useState), mais il vous permet de déplacer la logique de mise à jour de l'état depuis les gestionnaires d'événements vers une seule fonction à l'extérieur de votre composant. Apprenez-en davantage sur [comment choisir entre `useState` et `useReducer`](/learn/extracting-state-logic-into-a-reducer#comparing-usestate-and-usereducer).
```

  // ...
```

> Les noms des types d'actions sont locales à votre composant. [Chaque action décrit une seule interaction, même si ça amène à modifier plusieurs fois la donnée.](/learn/extracting-state-logic-into-a-reducer#writing-reducers-well) La forme de l'état est arbitraire, mais ce sera généralement un objet ou un tableau.

Reviewer: ```suggestion
Les noms des types d'actions sont locaux à votre composant. [Chaque action décrit une seule interaction, même si ça amène à modifier plusieurs fois la donnée](/learn/extracting-state-logic-into-a-reducer#writing-reducers-well). L'état peut avoir n'importe quelle forme, mais ce sera généralement un objet ou un tableau.
```

"arbitraire" en FR n'a pas le même sens, cf. Glossaire.

Lisez [Extraire la logique d'état dans un réducteur](/learn/extracting-state-logic-into-a-reducer) pour en apprendre davantage.

<Pitfall>

L'état est en lecture seule. Ne modifiez pas les objets ou les tableaux dans l'état :

```js {4,5}
function reducer(state, action) {
  switch (action.type) {
    case 'incremented_age': {

>       // 🚩 Ne modifiez pas l'objet dans l'état de cette façon :

Reviewer: ```suggestion
      // 🚩 Ne modifiez pas l’objet dans l’état de cette façon :
```

      state.age = state.age + 1;
      return state;
    }
```

> À la place, renvoyez toujours de nouveaux objets depuis votre réducteur :

Reviewer: ```suggestion
Au lieu de ça, renvoyez toujours de nouveaux objets depuis votre réducteur :
```

cf. Glossaire

        onChange={handleInputChange}
      />
      <button onClick={handleButtonClick}>

>         Incrémenter l'âge

Reviewer: ```suggestion
        Incrémenter l’âge
```

<Solution />

#### Liste de tâches (tableau) {/*todo-list-array*/}

> Dans cet exemple, le réducteur gère un tableau de tâches. Le tableau doit être mis à jour [sans modification](/learn/updating-arrays-in-state).

Reviewer: ```suggestion
Dans cet exemple, le réducteur gère un tableau de tâches. Le tableau doit être mis à jour [en respectant l'immutabilité](/learn/updating-arrays-in-state).
```

C'était vraiment trop déroutant sinon 😉

<Solution />

#### Écrire de la logique concise avec Immer {/*writing-concise-update-logic-with-immer*/}

> S'il est fastidieux de mettre à jour les tableaux et les objets sans les modifier, vous pouvez utiliser une bibliothèque comme [Immer](https://github.com/immerjs/use-immer#useimmerreducer) pour réduire le code répétitif. Immer vous permet d'écrire du code plus concis comme si vous changiez les objets, mais sous le capot, il effectue des mises à jour immuables :

Reviewer: ```suggestion
S'il est fastidieux de mettre à jour les tableaux et les objets sans les modifier en place, vous pouvez utiliser une bibliothèque comme [Immer](https://github.com/immerjs/use-immer#useimmerreducer) pour réduire le code répétitif. Immer vous permet d'écrire du code plus concis comme si vous modifiiez directement les objets, mais sous le capot, il effectue des mises à jour respectant l'immutabilité :
```

  // ...
```

> Bien que le résultat de `createInitialState(username)` soit seulement utilisé pour le premier rendu, vous continuez d'appeler cette fonction à chaque rendu. C'est du gâchis si elle crée de grands tableaux ou effectue des calculs gourmands.

Reviewer: ```suggestion
Bien que le résultat de `createInitialState(username)` soit seulement utilisé pour le premier rendu, vous continuez d'appeler cette fonction à chaque rendu. C'est du gâchis si elle crée de grands tableaux ou effectue des calculs coûteux.
```

  // ...
```

> Remarquez que vous passez `createInitialState`, qui est *elle-même une fonction*, et non `createInitialState()`, qui est le résultat de son exécution. De cette façon, l'état initial n'est pas recréé après l'initialisation.

Reviewer: ```suggestion
Remarquez que vous passez `createInitialState`, à savoir *la fonction elle-même*, et non `createInitialState()`, qui est le résultat de son exécution. De cette façon, l'état initial n'est pas recréé après l'initialisation.
```

Dans l'exemple ci-dessus, `createInitialState` prend un argument `username`. Si votre fonction d'initialisation n'a besoin d'aucune information pour calculer l'état initial, vous pouvez passez `null` comme second argument à `useReducer`.

<Recipes titleText="La différence entre passer une fonction d'initialisation et un état initial directement" titleId="examples-initializer">

#### Passer la fonction d'initialisation {/*passing-the-initializer-function*/}

> Cet exemple passe la fonction d'initialisation, la fonction `createInitialState` ne s'exécute que durant l'initialisation. Elle ne s'exécute pas lorsque le composant fait de nouveau son rendu, comme lorsque vous tapez dans le champ de saisie.

Reviewer: ```suggestion
Cet exemple passe la fonction d'initialisation, la fonction `createInitialState` ne s'exécute donc que durant l'initialisation. Elle ne s'exécute pas lorsque le composant fait de nouveau son rendu, comme lorsque vous tapez dans le champ de saisie.
```

<Solution />

#### Passer l'état initial directement {/*passing-the-initial-state-directly*/}

> Cet exemple **ne passe pas** la fonction d'initialisation, donc la fonction `createInitialState` s'exécute à chaque rendu, comme lorsque vous tapez dans le champ de saisie. Il n'y a pas de différence dans le comportement, mais ce code est moins efficace.

Reviewer: ```suggestion
Cet exemple **ne passe pas** la fonction d'initialisation, donc la fonction `createInitialState` s'exécute à chaque rendu, comme lorsque vous tapez dans le champ de saisie. Il n'y a pas de différence perceptible dans le comportement, mais ce code est moins efficace.
```

---

## Dépannage {/*troubleshooting*/}

> ### J'ai dispatché une action, mais les journaux m'affichent l'ancienne valeur {/*ive-dispatched-an-action-but-logging-gives-me-the-old-state-value*/}

Reviewer: ```suggestion
### J'ai dispatché une action, mais la console m'affiche l'ancienne valeur {/*ive-dispatched-an-action-but-logging-gives-me-the-old-state-value*/}
```

On logue en général dans la console, et "journaux" n'est décidément pas assez répandu…

Appeler la fonction `dispatch` **ne change pas l'état dans le code qui est en train de s'exécuter** :

```js {4,5,8}
function handleClick() {
  console.log(state.age);  // 42

  dispatch({ type: 'incremented_age' }); // Demande un nouveau rendu avec 43
  console.log(state.age);  // Toujours 42 !

  setTimeout(() => {
    console.log(state.age); // 42 ici aussi !
  }, 5000);
}
```

> C'est parce que [l'état se comporte comme un instantané](/learn/state-as-a-snapshot). Mettre à jour un état nécessite un nouveau rendu avec sa nouvelle valeur, mais n'affecte pas la variable JavaScript `state` dans votre gestionnaire d'événement qui est en cours d'exécution.

Reviewer: ```suggestion
C'est parce que [l'état se comporte comme un instantané](/learn/state-as-a-snapshot). Mettre à jour un état planifie un nouveau rendu avec sa nouvelle valeur, mais n'affecte pas la variable JavaScript `state` dans votre gestionnaire d'événement en cours d'exécution.
```

---

### J'ai dispatché une action, mais l'écran ne se met pas à jour {/*ive-dispatched-an-action-but-the-screen-doesnt-update*/}

React **ignorera votre mise à jour si le prochain état est égal à l'état précédent**, déterminé avec une comparaison [`Object.is`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/Object/is). C'est généralement ce qui arrive quand vous changez l'objet ou le tableau directement dans l'état :

```js {4-5,9-10}
function reducer(state, action) {
  switch (action.type) {
    case 'incremented_age': {

>       // 🚩 Incorrect : modification de l'objet existant

Reviewer: ```suggestion
      // 🚩 Incorrect : modification de l’objet existant
```

      state.age++;
      return state;
    }
    case 'changed_name': {

>       // 🚩 Incorrect : modification de l'objet existant

Reviewer: ```suggestion
      // 🚩 Incorrect : modification de l’objet existant
```

}
```

> Vous avez modifié un objet existant de `state` puis l'avez renvoyé, React a ainsi ignoré la mise à jour. Pour corriger ça, vous devez toujours vous assurer que vous [mettez à jour les objets dans l'état](/learn/updating-objects-in-state) et [mettez à jour les tableaux dans l'état](/learn/updating-arrays-in-state) plutôt que de les modifier :

Reviewer: ```suggestion
Vous avez modifié un objet existant de `state` puis l'avez renvoyé, React a ainsi ignoré la mise à jour. Pour corriger ça, vous devez toujours vous assurer que vous [mettez à jour les objets dans l'état](/learn/updating-objects-in-state) et [mettez à jour les tableaux dans l'état](/learn/updating-arrays-in-state) plutôt que de les modifier en place :
```

```js {4-8,11-15}
function reducer(state, action) {
  switch (action.type) {
    case 'incremented_age': {

>       // ✅ Correct : création d'un nouvel objet

Reviewer: ```suggestion
      // ✅ Correct : création d’un nouvel objet
```

      return {
        ...state,
        age: state.age + 1
      };
    }
    case 'changed_name': {

>       // ✅ Correct : création d'un nouvel objet

Reviewer: ```suggestion
      // ✅ Correct : création d’un nouvel objet
```

---

### Une partie de l'état de mon réducteur devient undefined après le dispatch {/*a-part-of-my-reducer-state-becomes-undefined-after-dispatching*/}

> Assurez-vous que chaque branche `case` **copie tous les champs existants** lorsqu'il renvoie le nouvel état :

Reviewer: ```suggestion
Assurez-vous que chaque branche `case` **copie tous les champs existants** lorsqu'elle renvoie le nouvel état :
```

```js {5}
function reducer(state, action) {
  switch (action.type) {
    case 'incremented_age': {
      return {

>         ...state, // N'oubliez pas ceci !

Reviewer: ```suggestion
        ...state, // N’oubliez pas ceci !
```

        age: state.age + 1
      };
    }
    // ...
```

Sans le `...state` ci-dessus, le prochain état renvoyé ne contiendrait que le champ `age` et rien d'autre.

---

### Tout l'état de mon réducteur devient undefined après le dispatch {/*my-entire-reducer-state-becomes-undefined-after-dispatching*/}

> Si votre état devient `undefined` de manière imprévue, c'est que vous avez probablement oublié de `return` l'état dans l'un de vos cas, ou que le type d'action ne correspond à aucune des instructions `case`. Pour comprendre pourquoi, lancez une erreur à l'extérieur du `switch` :

Reviewer: ```suggestion
Si votre état devient `undefined` de manière imprévue, c'est que vous avez probablement oublié de `return` l'état dans l'un de vos cas, ou que le type d'action ne correspond à aucune des instructions `case`. Pour comprendre pourquoi, levez une erreur à l'extérieur du `switch` :
```

}
```

Vous pouvez également utiliser un vérificateur de type statique comme TypeScript pour détecter ces erreurs.

---

> ### J'obtiens l'erreur « Too many re-renders » {/*im-getting-an-error-too-many-re-renders*/}

Reviewer: ```suggestion
### J'obtiens l'erreur “Too many re-renders” {/*im-getting-an-error-too-many-re-renders*/}

*(« Trop de rendus successifs », NdT)*
```

> Vous pouvez rencontrer l'erreur indiquant `Too many re-renders. React limits the number of renders to prevent an infinite loop.` (*Trop de nouveaux rendus. React limite le nombre de rendus pour éviter des boucles infinies*, NdT). Ça signifie généralement que vous dispatchez une action inconditionnellement *pendant un rendu* et ainsi votre composant entre dans une boucle : rendu, dispatch (qui occasionne un rendu), rendu, dispatch (qui occasionne un rendu), et ainsi de suite. La cause en est très souvent une erreur dans la spécification d'un gestionnaire d'événement :

Reviewer: ```suggestion
Vous pouvez rencontrer l'erreur indiquant `Too many re-renders. React limits the number of renders to prevent an infinite loop.` *(« Trop de rendus successifs. React limite le nombre de rendus pour éviter des boucles infinies », NdT)* Ça signifie généralement que vous dispatchez une action inconditionnellement *pendant un rendu* et ainsi votre composant entre dans une boucle : rendu, dispatch (qui occasionne un rendu), rendu, dispatch (qui occasionne un rendu), et ainsi de suite. Ça vient le plus souvent d'une erreur dans la spécification d'un gestionnaire d'événement :
```

```js {1-2}
// 🚩 Incorrect : appelle le gestionnaire pendant le rendu

> return <button onClick={handleClick()}>Click me</button>

Reviewer: ```suggestion
return <button onClick={handleClick()}>Cliquez ici</button>
```

(À reporter sur le reste de l'exemple)

> // ✅ Correct : passe le gestionnaire d'événement

Reviewer: ```suggestion
// ✅ Correct : passe le gestionnaire d’événement
```

return <button onClick={handleClick}>Click me</button>

// ✅ Correct : passe une fonction en ligne
return <button onClick={(e) => handleClick(e)}>Click me</button>
```

> Si vous ne trouvez pas la cause de cette erreur, cliquez sur la flèche à côté de l'erreur dans la console et parcourez la pile JavaScript pour trouver l'appel à la fonction `dispatch` responsable de l'erreur.

Reviewer: ```suggestion
Si vous ne trouvez pas la cause de cette erreur, cliquez sur la flèche à côté de l'erreur dans la console et parcourez la pile d'appels JavaScript pour trouver l'appel à la fonction `dispatch` responsable de l'erreur.
```

---

### Mon réducteur ou ma fonction d'initialisation s'exécute deux fois {/*my-reducer-or-initializer-function-runs-twice*/}

> Dans le [Mode Strict](/reference/react/StrictMode), React appellera votre réducteur et votre fonction d'initialisation deux fois. Ça ne devrait pas casser votre code.

Reviewer: ```suggestion
En [mode strict](/reference/react/StrictMode), React appellera votre réducteur et votre fonction d'initialisation deux fois. Ça ne devrait pas casser votre code.
```

Ce comportement **spécifique au développement** vous aide à [garder les composants purs](/learn/keeping-components-pure). React utilise le résultat de l'un des appels et ignore l'autre. Tant que votre composant, votre fonction d'initialisation et votre réducteur sont purs, ça ne devrait pas affecter votre logique. Si toutefois ils sont malencontreusement impurs, ça vous permettra de détecter les erreurs.

> Par exemple, cette fonction de réduction impure modifie un tableau dans l'état :

Reviewer: ```suggestion
Par exemple, cette fonction de réduction impure modifie directement un tableau dans l'état :
```

```js {4-6}
function reducer(state, action) {
  switch (action.type) {
    case 'added_todo': {

>       // 🚩 Erreur : modification de l'état

Reviewer: ```suggestion
      // 🚩 Erreur : modification de l’état
```

}
```

> Comme React appelle deux fois votre fonction de réduction, vous verrez que la liste a été ajoutée deux fois, vous saurez donc qu'il y a un problème. Dans cet exemple, vous pouvez le corriger [en remplaçant le tableau plutôt qu'en le modifiant](/learn/updating-arrays-in-state#adding-to-an-array) :

Reviewer: ```suggestion
Comme React appelle deux fois votre fonction de réduction, vous verrez que la liste a été ajoutée deux fois, vous saurez donc qu'il y a un problème. Dans cet exemple, vous pouvez le corriger [en remplaçant le tableau plutôt que de le modifier en place](/learn/updating-arrays-in-state#adding-to-an-array) :
```

```js {4-11}
function reducer(state, action) {
  switch (action.type) {
    case 'added_todo': {

>       // ✅ Correct : remplacer avec un nouvel état

Reviewer: ```suggestion
      // ✅ Correct : remplacement par un nouvel état
```

Cohérence de forme avec la version incorrecte (substantif, pas infinitif)

}
```

> Maintenant que cette fonction de réduction est pure, l'appeler une fois de plus ne change pas le comportement. C'est pourquoi React vous aide à trouver l'erreur en l'appelant deux fois. **Seuls les composants, les fonctions d'initialisations et les réducteurs doivent être purs.** Les gestionnaires d'événements n'ont pas besoin de l'être, React ne les appellera donc jamais deux fois.

Reviewer: ```suggestion
Maintenant que cette fonction de réduction est pure, l'appeler une fois de plus ne change pas le comportement. C'est pourquoi React vous aide à détecter des erreurs en l'appelant deux fois. **Seuls les composants, les fonctions d'initialisation et les réducteurs doivent être purs.** Les gestionnaires d'événements n'ont pas besoin de l'être, React ne les appellera donc jamais deux fois.
```
